# 每天一个小知识点

## 第二十五天

---

## let 是否存在变量提升

根据 es 文档,let 确实存在变量提升。只不过由于暂时性死区的限制,不能在 let 声明变量之前使用变量  
1、var 声明的【创建、初始化、赋值过程】

```js
function fn() {
  var x = 1;
  var y = 2;
}
fn();
```

- 进入 fn,为 fn 创建一个环境
- 找到 fn 中所有用 var 声明的变量,在这个环境中创建这些变量
- 将这些变量初始化为 undefined
- 执行代码
- 将 x 的值赋为 1
- 将 y 的值赋为 2
  var 声明会在代码执行之前创建变量并初始化为 undefinded
  2、function 声明的【创建、初始化、赋值】过程

```js
fn();
function fn() {
  console.log(1);
}
```

- 找到所有用 function 声明的变量,在环境中创建这些变量
- 将这些变量初始化并赋值
- 开始执行函数
  function 声明会在代码执行执行之前创建并赋值
  3、let 声明【创建、初始化、赋值】过程

```js
let x = 1;
x = 2;
```

- 找到所有 let 声明的变量,在环境中创建这些变量
- 开始执行代码(此时还没有初始化)
- 执行`x = 1`,将 x 初始化为 1(这并不是赋值,如果代码是 let x,则将 x 初始化为 undefined)
- 执行`x = 2`,对 x 进行赋值

```js
let x = 'global';
{
  console.log(x); //  Uncaught ReferenceError: x is not defined
  let x = 1;
}
```

报错原因有两个:  
1、`console.log(x)`中的 x 指的是后边的 x,而不是全局 x  
2、执行打印函数时,let 还没有初始化,所以不能使用(暂时性死区)
总结:

- let 的【创建】过程被提升了,但是初始化没有提升
- var【创建、初始化】都被提升了
- function【创建、初始化、赋值】都被提升了
- const 只有【创建、初始化】,没有赋值过程

## JS 原型链的理解，并且是否有创建实例时隐式原型**proto**赋值的特殊情况

```js
Object.prototype.__proto__ === null; // true
Object.__proto__ === Function.prototype; // true
Function.__proto__ === Function.prototype; //  true
Function.prototype.__proto__ === Object.prototype; // true
Function.prototype === Function.__proto__; // true
```

所有的构造函数都可以看作 Function 的实例对象，Function 自己也是 Function 构造的  
Object 要 Function 的 prototype 去创建，但 Function 的 prototype 要 Object 去创建
![原型链](/docs/images/EveryDayKnowledge/prototype.jpg)

## 执行结果

```js
// 代码一
new Promise((resolve, reject) => {
  console.log('外部promise'); // 1
  resolve();
})
  .then(() => {
    console.log('外部第一个then'); // 2
    return new Promise((resolve, reject) => {
      console.log('内部promise'); // 3
      resolve();
    })
      .then(() => {
        console.log('内部第一个then'); // 4
      })
      .then(() => {
        console.log('内部第二个then'); // 5
      });
  })
  .then(() => {
    console.log('外部第二个then'); // 6
  });
// 如果promise 有return，就是等这个promise里面的所有then都执行完，返回一个promise才会执行后面的任务。
// 代码二
new Promise((resolve, reject) => {
  console.log('外部promise'); //1
  resolve();
})
  .then(() => {
    console.log('外部第一个then'); // 2
    new Promise((resolve, reject) => {
      console.log('内部promise'); // 3
      resolve();
    })
      .then(() => {
        console.log('内部第一个then'); // 4
      })
      .then(() => {
        console.log('内部第二个then'); // 6
      });
  })
  .then(() => {
    console.log('外部第二个then'); // 5
  });
// 如果直接new promise,那就是任务放到事件队列里面，先执行对级的then，再之后后面同级的then
```
